---
title: "R Notebook"
output: html_notebook
---
---
title: "R Notebook"
output: html_notebook
---
```{r}
library(tidyverse)
library(here)
library(Biostrings)
library(stringr)
library(rvest)
library(stringi)
sessionInfo()
install.packages("stringr")
install.packages("rvest")
```
```{r}
# A list for both targeted and RNASeq data
rds_output <- dir(path = ".", pattern = "\\.r.rds", full.names = FALSE, recursive = TRUE)
rds_output
targeted <- dir(path = ".", pattern = "\\.t.rds", full.names = FALSE, recursive = TRUE)
targeted
```
# The columns are not the same name .: renamed targeted so able to compare the two 
```{r}
targeted_1 <- readRDS("R502-12_S10__treg_R502_end.t.rds")
targeted_1
RNAseq_1 <- readRDS('R502-12_S10__treg_R502_end.r.rds')
t1 <- readRDS('treg_R518_start t.rds')
```
# For one sample to test: In order to get the length of the amino acid sequences between targeted and RNA_seq so able to compare. 
```{r}
targeted_1%>%
    rowwise()%>%
    mutate(frame1=Biostrings::translate(DNAString(as.character(nucleotide))%>%subseq(.,1), if.fuzzy.codon="X")%>%as.character(),   #this is to obtain all 3 reading frames
           frame2=Biostrings::translate(DNAString(as.character(nucleotide))%>%subseq(.,2), if.fuzzy.codon="X")%>%as.character(),
           frame3= Biostrings::translate(DNAString(as.character(nucleotide))%>%subseq(.,3), if.fuzzy.codon="X")%>%as.character())%>%
    rowwise()%>%
    mutate_at(vars(matches("frame.$")), list(whereStartAA=~ifelse(!is.na(.), stri_locate_first_fixed(pattern=aminoAcid, str =.))))%>%  #here you find aminoAcid in these 3 frames
    mutate(final_start=(min(c(frame1_whereStartAA, frame2_whereStartAA, frame3_whereStartAA), na.rm=T)-1)*3+1)%>%  #and identify which one is the apropriate one; get nucleotide locations from aminoacid location
    mutate( fr_cdr3_nt= ifelse(!is.na(aminoAcid), str_sub(nucleotide, start=final_start,  #obtain the nucleotide sequence
                                                       end=final_start+nchar(aminoAcid)*3-1)  ,NA),
            fr_check=ifelse(is.na(fr_cdr3_nt),NA_character_,Biostrings::translate(DNAString(fr_cdr3_nt))%>%as.character())) #and translate it back yo aminoacids, to compare with the expected one
```

# Loop so it does it for all the samples.
```{r}
for (i in targeted){
   sample_name <- sub("_TCRB.tsv", "", i)
  targeted_sample <- read_tsv(i)
a <- targeted_sample %>%
  rowwise()%>%
    mutate(frame1=Biostrings::translate(DNAString(as.character(nucleotide))%>%subseq(.,1), if.fuzzy.codon="X")%>%as.character(),   
           frame2=Biostrings::translate(DNAString(as.character(nucleotide))%>%subseq(.,2), if.fuzzy.codon="X")%>%as.character(),
           frame3= Biostrings::translate(DNAString(as.character(nucleotide))%>%subseq(.,3), if.fuzzy.codon="X")%>%as.character()) %>% 
rowwise()%>% 
    mutate_at(vars(matches("frame.$")), list(whereStartAA=~ifelse(!is.na(.), stri_locate_first_fixed(pattern=aminoAcid, str =.))))%>%  
    mutate(final_start=(min(c(frame1_whereStartAA, frame2_whereStartAA, frame3_whereStartAA), na.rm=T)-1)*3+1)%>%  
    mutate( nSeqCDR3= ifelse(!is.na(aminoAcid), str_sub(nucleotide, start=final_start, 
                                                       end=final_start+nchar(aminoAcid)*3-1)  ,NA),
            fr_check=ifelse(is.na(nSeqCDR3),NA_character_,Biostrings::translate(DNAString(nSeqCDR3))%>%as.character())) 
saveRDS(a,paste((sample_name),'t.rds'))
}
```

# Trial to see how join works, whether it is doing the right thing before writing the script 
```{r}

left_Joint <- left_join(RNAseq_1, targeted_1, by = "aaSeqCDR3", "aminaAcid", copy = F, suffix = c(".r",".t")) # Shows all the amino acids of RNASeq and NA if not present in aaSeqCDR3 of targeted. 
Inner_Joint <- inner_join(RNAseq_1, targeted_1, by = "aaSeqCDR3", copy = F, suffix = c(".r", ".t")) # Shows all the amino acid sequences that were similar .: only 85
semi_Joint <- semi_join(RNAseq_1, targeted_1, by = "nSeqCDR3", copy = F, suffix = c(".r", ".t"))
anti_join <- anti_join(RNAseq_1, targeted_1, by = "nSeqCDR3", copy = F, suffix = c(".r", ".t"))
anti_join1 <- anti_join(targeted_1, RNAseq_1, by = "nSeqCDR3", copy = F, suffix = c(".r", ".t"))
full_joint <- full_join(RNAseq_1, targeted_1, by = c("aaSeqCDR3" = "aminoAcid"), copy = F, suffix = c(".r", ".t"))
# To see how many aaSeqCDR3 sequences overlap .: drop_na, 85 out of ~2000 RNAseq amino acid sequences
Joint1 %>% drop_na(nSeqCDR3.t) %>% 
select(aaSeqCDR3, nSeqCDR3.r, fam_v, fam_j, cloneCount, samplename.1, nSeqCDR3.t, vFamilyName, jFamilyName, "count (templates/reads)") # This gives all the ones that matched = 85 same as inner join
```
```{r}
# One sample to test whether this works so then can use in the loop
# Changing targeted sequencing column fam_v and fam_j to be similar format to RNAseq. 
targeted_1$vFamilyName <- gsub("C","",targeted_1$vFamilyName) # Remove the C from TCRBV
targeted_1$jFamilyName <- gsub("C","",targeted_1$jFamilyName)
targeted_1$vFamilyName <- gsub("TRBV0","TRBV", targeted_1$vFamilyName) # Remove the zero from one digit numbers= TRBV01 to TRBV1
targeted_1$jFamilyName <- gsub("TRBV0","TRBV", targeted_1$vFamilyName)
```

```{r}
# step 1: change the column in targeted of family to be the same as RNASeq
# step 3: inner join the columns for both samples
# step 4: Save the outputs

```

```{r}
output <- dir(path = ".", pattern = "\\.rds$", full.names = FALSE, recursive = TRUE)
output
targeted_rds <- dir(path = ".", pattern = "\\.t.rds$", full.names = FALSE, recursive = TRUE)
targeted_rds
RNAseq_rds <- dir(path = ".", pattern = "\\.r.rds$", full.names = FALSE, recursive = TRUE)
readRDS(RNAseq_rds)

for (i in output){
  j <- sub("$t.rds", "$r.rds", i)
  h <- sub("$r.rds", "$t.rds", i)
  samplename <- sub("[r-t].rds", "", i)
  targeted <- readRDS(targeted_rds)
  RNASeq <- readRDS(RNASeq_RDS)
  targeted$vFamilyName <- gsub("C","",targeted$vFamilyName)
  targeted$jFamilyName <- gsub("C","",targeted$jFamilyName) 
  targeted$vFamilyName <- gsub("TRBV0","TRBV", targeted$vFamilyName)
  targeted$jFamilyName <- gsub("TRBV0","TRBV", targeted$vFamilyName)
  Joint <- inner_join(RNAseq, targeted, by = "nSeqCDR3", copy = T, suffix = c(".r", ".t"))
 Joint_1 <- Joint %>% 
    select(aaSeqCDR3, nSeqCDR3.r, fam_v, fam_j, cloneCount, samplename.1, nSeqCDR3.t, vFamilyName, jFamilyName, "count (templates/reads)", "fr_check")
  Joint_1$samplename <- samplename
  saveRDS(Joint_1, paste((h), paste(j), '.rds'))
} 
```
```{r}
for (i in output){
  j <- sub(".t.rds$", ".r.rds", i)
  h <- sub(".r.rds$", ".t.rds", i)
  samplename <- sub(".[r-t].rds$", "", i)
  targeted <- readRDS(h)
  RNASeq <- readRDS(j)
  targeted$vFamilyName <- gsub("C","",targeted$vFamilyName)
  targeted$jFamilyName <- gsub("C","",targeted$jFamilyName) 
  targeted$vFamilyName <- gsub("TRBV0","TRBV", targeted$vFamilyName)
  targeted$jFamilyName <- gsub("TRBJ0","TRBJ", targeted$jFamilyName)
  Joint <- left_join(RNASeq,targeted, by = "nSeqCDR3", copy = T, suffix = c(".r", ".t")) %>% 
    select(nSeqCDR3, aaSeqCDR3, cloneCount, fam_v, fam_j, fr_check, "count (templates/reads)", vFamilyName, jFamilyName)
  
  Joint$samplename <- samplename
} 
```
# using the anti_join to obtain the sequences that did not overlap between the two
```{r}
for (i in output){
  j <- sub("t.rds", "r.rds", i)
  h <- sub("r.rds", "t.rds", i)
  samplename <- sub("[r-t].rds", "", i)
  targeted <- readRDS(h)
  RNASeq <- readRDS(j)
  Joint <- anti_join(RNASeq, targeted, by = "nSeqCDR3", copy = T, suffix = c(".r", ".t"))
  Joint_1 <- Joint %>% 
    select(aaSeqCDR3, nSeqCDR3, fam_v, fam_j, cloneCount, samplename.1)
  saveRDS(Joint_1, paste((samplename), 'ANTI_rnaSeq.rds'))
} 
```

```{r}
for (i in output){
  j <- sub("t.rds", "r.rds", i)
  h <- sub("r.rds", "t.rds", i)
  samplename <- sub("[r-t].rds", "", i)
  targeted <- readRDS(h)
  RNASeq <- readRDS(j)
  targeted$vFamilyName <- gsub("C","",targeted$vFamilyName)
  targeted$jFamilyName <- gsub("C","",targeted$jFamilyName) 
  targeted$vFamilyName <- gsub("TRBV0","TRBV", targeted$vFamilyName)
  targeted$jFamilyName <- gsub("TRBJ0","TRBJ", targeted$jFamilyName)
  
  Joint <- anti_join(targeted,RNASeq, by = "nSeqCDR3", copy = T, suffix = c(".r", ".t"))
  Joint_1 <- Joint %>% 
     select(nSeqCDR3, fr_check, "count (templates/reads)", vFamilyName, jFamilyName)
  saveRDS(Joint_1, paste((samplename), 'ANTI_targeted.rds'))
} 
```

# Finding the number of sequences the samples started with
```{r}
for (i in targeted){ 
a <-readRDS(i)
str(a)
}
for (i in targeted){ 
a <-readRDS(i)
a1<-  filter(a, sequenceStatus == "In") 
str(a1)
}
```


# Obtain the sequences that did not overlap when the results were filtered to not contain IN samples 
```{r}
for (i in output){
  j <- sub("t.rds", "r.rds", i)
  h <- sub("r.rds", "t.rds", i)
  samplename <- sub("[r-t].rds", "", i)
  targeted <- readRDS(h)
  RNASeq <- readRDS(j)
  targeted$vFamilyName <- gsub("C","",targeted$vFamilyName)
  targeted$jFamilyName <- gsub("C","",targeted$jFamilyName) 
  targeted$vFamilyName <- gsub("TRBV0","TRBV", targeted$vFamilyName)
  targeted$jFamilyName <- gsub("TRBJ0","TRBJ", targeted$jFamilyName)
  targeted_1 <- filter(targeted,sequenceStatus == "In")
  Joint <- anti_join(targeted_1,RNASeq, by = "nSeqCDR3", copy = T, suffix = c(".r", ".t"))
  Joint_1 <- Joint %>% 
     select(nSeqCDR3, fr_check, "count (templates/reads)", vFamilyName, jFamilyName)
  saveRDS(Joint_1, paste((samplename), 'ANTI_targeted_IN.rds'))
} 
```